0022-lnd-plan-of-attack.adoc

```diff
@@ -0,0 +1,448 @@
+
+One could conceptualize a single Lightning channel as a ledger in the way that it tracks ownership of who owns how many Bitcoins.
+Using a channel would be more correct than using a single node but bears other problems.
+In particular, payments in Lightning seldomly involve a single channel.
+Using the channel = Ledger model would work if there were not hops involved within a payment in Lightning.
+
+===== Why referring to the whole Lightning network as the Ledger
+
+It is conceptually similar to how we see other ledgers (Bitcoin, Ethereum).
+The difference is that it is a private ledger, hence not every party can see every transaction.
+I feel the concept of a "private" ledger is better suited here.
+With https://github.com/comit-network/comit-rs/issues/1862 landing, we can introduce ledgers like `lightning::Mainnet` etc.
+
+==== What does the protocol look like?
+
+The following diagrams assume that a negotiation protocol is in place and the parties agreed on a set of parameters.
+The protocol is following the "COMIT spirit":
+
+- readonly access to LND; COMIT is never in control of the user's funds
+- moving money is explicit; COMIT emits a dedicated `redeem` action for both protocol (even though it would not be necessary for the HAN-HALight one because Alice knows the secret from the very beginning so she could just use a regular invoice)
+
+===== HAN-HALight swap
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/3e474ff2/assets/0022-halight-eth-ln-swap.puml&fmt=svg[HAN-HALight]
+
+===== HALight-HAN swap
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/3e474ff2/assets/0022-halight-ln-eth-swap.puml&fmt=svg[HAlight-HAN]
+
+==== Which parameters do we need to exchange for the protocol?
+
+To make a payment in Lightning we need to know:
+
+- the recipient pubkey
+- the amount
+- the secret hash
+- the final-cltv-delta
+
+The payment can then by made (using lncli):
+
+[source=bash]
+----
+lncli sendpayment \
+    --dest=0346093cc4b9010fa3885df8dfcb6015bc2190bc9f46f5935a48df0417eeb7667e \
+    --amt=1000 \
+    --payment_hash=ec4916dd28fc4c10d78e287ca5d9cc51ee1ae73cbfde08c6b37324cbfaac8bc5 \
+    --final_cltv_delta=40
+----
+
+==== Which LND APIs do we need to target?
+
+COMIT only needs read-access to LND.
+We need APIs for:
+
+* Monitor an invoice
+* Monitor an outgoing payment
+
+===== Watching an invoice
+
+We need to be able to differentiate between the following states:
+
+* Hold invoice added to LND but not paid yet (see <<HAN-HALight swap>>, cnd needs to be able detect that the invoice was added to emit the next action)
+* Invoice has been paid (analogous to HTLC has been deployed)
+* Invoice has been settled (analogous to HTLC has been redeemed)
+* Invoice has been rejected (this does not exist on Layer 1, it would be analogous for removing the HTLC from the blockchain again)
+* Invoice has timed out
```

Comment:
Can this happen if there is a pending payment on such invoice? How much do we control that?